### ISSUE TITLE
geckodriver trying to launch firefox directory from $PATH

### ISSUE THREAD

**Issue by maclarke** [mozilla/geckodriver#1288]
## System

* Version: 0.20.1
* Platform: Linux
* Firefox: 59.0.3
* Selenium: 3.12.0

## Testcase

I was having an issue getting my selenium tests to run and eventually got down to running geckodriver manually and send it commands with curl. With this I eventually figured out that as geckodriver is running through $PATH it just checks the name but not the type.

In my case geckodriver encountered $HOME/local/bin/firefox (a directory) and tried to launch it - which failed

```
$ PATH="$HOME/local/bin:$HOME/local/bin/firefox" geckodriver --log trace        
1527275764768   geckodriver     INFO    geckodriver 0.20.1
1527275764768   webdriver::httpapi      DEBUG   Creating routes
1527275764772   geckodriver     INFO    Listening on 127.0.0.1:4444
1527275825841   webdriver::server       DEBUG   -> POST /session {"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}
1527275825843   mozrunner::runner       INFO    Running command: "/home/mclarke/local/bin/firefox" "-marionette" "-profile" "/tmp/rust_mozprofile.OpvFg9s8Wuby"
1527275825844   webdriver::server       DEBUG   <- 500 Internal Server Error {"value":{"error":"session not created","message":"Failed to start browser /home/mclarke/local/bin/firefox: permission denied","stacktrace":""}}
```

When I reverse the paths in $PATH, it hits the actual executable first and things go as expected

```
$ PATH="$HOME/local/bin/firefox:$HOME/local/bin" geckodriver --log trace
1527276115910   geckodriver     INFO    geckodriver 0.20.1
1527276115910   webdriver::httpapi      DEBUG   Creating routes
1527276115914   geckodriver     INFO    Listening on 127.0.0.1:4444
1527276117793   webdriver::server       DEBUG   -> POST /session {"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}
1527276117795   mozrunner::runner       INFO    Running command: "/home/mclarke/local/bin/firefox/firefox" "-marionette" "-profile" "/tmp/rust_mozprofile.DuDT6DIwvfgw"
1527276117795   geckodriver::marionette DEBUG   Waiting 60s to connect to browser
<SNIP>
1527276119564   Marionette      TRACE   0 <- [1,1,null,{"sessionId":"4ebe4844-3c62-4aa8-92c8-550f622d26c6","capabilities":{"browserName":"firefox","browserVersion":"59.0. ... 6,"moz:profile":"/tmp/rust_mozprofile.DuDT6DIwvfgw","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1527276119566   geckodriver::marionette TRACE   <- [1,1,null,{"sessionId":"4ebe4844-3c62-4aa8-92c8-550f622d26c6","capabilities":{"browserName":"firefox","browserVersion":"59.0.3","platformName":"linux","platformVersion":"4.9.0-6-amd64","pageLoadStrategy":"normal","acceptInsecureCerts":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"rotatable":false,"moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":15116,"moz:profile":"/tmp/rust_mozprofile.DuDT6DIwvfgw","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1527276119567   webdriver::server       DEBUG   <- 200 OK {"value": {"sessionId":"4ebe4844-3c62-4aa8-92c8-550f622d26c6","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"59.0.3","moz:accessibilityChecks":false,"moz:headless":false,"moz:processID":15116,"moz:profile":"/tmp/rust_mozprofile.DuDT6DIwvfgw","moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"linux","platformVersion":"4.9.0-6-amd64","rotatable":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000}}}}

```

 In both cases I am sending the following command with curl
```
$ curl -d  '{"acceptInsecureCerts":true,"browserName":"firefox","capabilities":{"desiredCapabilities":{"acceptInsecureCerts":true,"browserName":"firefox"}}}'    http://localhost:4444/session
```

## Context

My $HOME/local/bin directory is organized as follows:

```
 $ ls -lh ~/local/bin 
total 7.3M
drwxr-xr-x 9 mclarke mclarke 4.0K May 25 13:55 firefox
-rwxr-xr-x 1 mclarke mclarke 7.3M Apr  8 07:51 geckodriver

$ ls -lh ~/local/bin/firefox/firefox
-rwxr-xr-x 1 mclarke mclarke 184K Apr 27 17:14 /home/mclarke/local/bin/firefox/firefox
```

This was created by downloading and unpacking the geckodriver and firefox tarballs

```
$ wget https://github.com/mozilla/geckodriver/releases/download/v0.20.1/geckodriver-v0.20.1-linux64.tar.gz
$ tar -xzf geckodriver-v0.20.1-linux64.tar.gz -C ~/local/bin

$ wget "https://ftp.mozilla.org/pub/firefox/releases/59.0.3/linux-x86_64/en-US/firefox-59.0.3.tar.bz2"
$ tar -xjf firefox-59.0.3.tar.bz2 -C ~/local/bin/
```

**Comment by whimboo** [comment]
The problem here is that geckodriver only checks if the path is existent but not if it is a file and executable. As such it returns with the first found path with `firefox` in it, which then fails later. The underlying problem is in mozrunner. So I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1464995 to get this fixed.

For now just rename the `firefox` folder to something else like `firefox-release` or `firefox-61.0`, and it will also work.

**Comment by andreastt** [comment]
This is fixed in 0.21.0.
